### ISSUE TITLE
Gunicorn 20.1.0 introduced changing behavior with print_config

### ISSUE THREAD

**Issue by wjentner** [benoitc/gunicorn#2551]
We used `print_config = True` to dump the gunicorn configuration before the start of a server. This worked before gunicorn 20.1.0.
With the update, the configuration is being printed, but the process is then terminated.

**Comment by laggardkernel** [comment]
I don't think so. Reading from the source, `print_config` always works like this -- print the conf and quit. The behavior of `print_config` has not been changed since the introduction of this setting in 2019-05-31 13c5d72

What version of gunicorn did you use before upgrading to 20.1.0?

**Comment by wjentner** [comment]
The version before was 20.0.4, and it was working there. We are using a CI/CD pipeline with automatic dependency updates (renovate-bot) on feature branches and deploy everything into a testing environment on Kubernetes. The Pods kept crashing because the liveness probes (HTTP status) were not working since the container immediately exited after printing the config. On 20.0.4 and before the config was printed to stdout and the server started.

Config file:

```python
import multiprocessing

bind = "0.0.0.0:5000"
workers = multiprocessing.cpu_count() * 2 + 1
threads = multiprocessing.cpu_count() * 2 + 1

# doe not work with gunicorn v20.1.0 anymore
# print_config = True

accesslog = '-'

loglevel = 'debug'

# 2 minute timeout for worker because of long query times
timeout = 240
```

Start command:

```sh
gunicorn --config=gunicorn.conf.py server:app
```

**Comment by wjentner** [comment]
I did create a minimal setup here: https://github.com/wjentner/gunicorn-test
and also pushed some docker images with the two gunicorn version (20.0.4, 20.1.0). You can simply run these with `docker run ...` and you will see the different behavior.

**Comment by laggardkernel** [comment]
## Brief
`print_config` had not been introduced when Gunicorn 20.0.4 is released. This setting is ignored in 20.0.4. What you saw was not `print`ed, but logged.

Just comment out `print_config` in your `gunicorn.conf.py`.

## Detail (Original answer)
Sorry, I missed something. Commit 13c5d72, which introduces setting `print_config`, was created on 2019-05-31 but had not been merged when 20.0.4 was released. Setting `print_config` was just ignored in Gunicorn 20.0.4. The setting output you saw with gunicorn 20.0.4 was not `print`ed, it was **logged**.

Here're some calling tracings help me explain it,

In gunicorn 20.1.0

`WSGIApplication.run()` -> `Application.run()` -> `print(self.cfg)`, then `sys.exit()`

```python
class Application(BaseApplication):
    ...
    def run(self):
        if self.cfg.print_config:
            print(self.cfg)

        if self.cfg.print_config or self.cfg.check_config:
            try:
                self.load()
            except Exception:
                msg = "\nError while loading the application:\n"
                print(msg, file=sys.stderr)
                traceback.print_exc()
                sys.stderr.flush()
                sys.exit(1)
            sys.exit(0)
```

While, in gunicorn 20.0.4, `print_config` setting is not added,

`WSGIApplication.run()` -> `Application.run()` -> `BaseApplication.run()` -> `Arbiter.__init__()` -> `Arbiter.setup()` -> `Arbiter.log.debug()`, "print" your config and continue serve requests.

```python
class Arbiter(object):
    ...
    def setup(self, app):
        ...
        self.log.debug('Current configuration:\n{0}'.format(
            '\n'.join(
                '  {0}: {1}'.format(config, value.value)
                for config, value
                in sorted(self.cfg.settings.items(),
                          key=lambda setting: setting[1]))))
```

If you look carefully, the logged setting and printed setting are shown in different format.

<details>
<summary>
Logged settings in Gunicorn 20.0.4
</summary>

```
[97998] [DEBUG] Current configuration:
  config: gunicorn.conf.py
  bind: ['0.0.0.0:5000']
  backlog: 2048
  workers: 17
  worker_class: sync
  threads: 17
  worker_connections: 1000
  max_requests: 0
  max_requests_jitter: 0
  timeout: 240
  graceful_timeout: 30
  keepalive: 2
  limit_request_line: 4094
  limit_request_fields: 100
  limit_request_field_size: 8190
  reload: False
  reload_engine: auto
  reload_extra_files: []
  spew: False
  check_config: False
  preload_app: False
  sendfile: None
  reuse_port: False
  chdir: /Users/username/Workspace/tmp/test_gunicorn/test_print
  daemon: False
  raw_env: []
  pidfile: None
  worker_tmp_dir: None
  user: 501
  group: 20
  umask: 0
  initgroups: False
  tmp_upload_dir: None
  secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
  forwarded_allow_ips: ['127.0.0.1']
  accesslog: -
  disable_redirect_access_to_syslog: False
  access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
  errorlog: -
  loglevel: debug
  capture_output: False
  logger_class: gunicorn.glogging.Logger
  logconfig: None
  logconfig_dict: {}
  syslog_addr: unix:///var/run/syslog
  syslog: False
  syslog_prefix: None
  syslog_facility: user
  enable_stdio_inheritance: False
  statsd_host: None
  dogstatsd_tags:
  statsd_prefix:
  proc_name: None
  default_proc_name: app:app
  pythonpath: None
  paste: None
  on_starting: <function OnStarting.on_starting at 0x1100325e0>
  on_reload: <function OnReload.on_reload at 0x110032700>
  when_ready: <function WhenReady.when_ready at 0x110032820>
  pre_fork: <function Prefork.pre_fork at 0x110032940>
  post_fork: <function Postfork.post_fork at 0x110032a60>
  post_worker_init: <function PostWorkerInit.post_worker_init at 0x110032b80>
  worker_int: <function WorkerInt.worker_int at 0x110032ca0>
  worker_abort: <function WorkerAbort.worker_abort at 0x110032dc0>
  pre_exec: <function PreExec.pre_exec at 0x110032ee0>
  pre_request: <function PreRequest.pre_request at 0x110048040>
  post_request: <function PostRequest.post_request at 0x1100480d0>
  child_exit: <function ChildExit.child_exit at 0x1100481f0>
  worker_exit: <function WorkerExit.worker_exit at 0x110048310>
  nworkers_changed: <function NumWorkersChanged.nworkers_changed at 0x110048430>
  on_exit: <function OnExit.on_exit at 0x110048550>
  proxy_protocol: False
  proxy_allow_ips: ['127.0.0.1']
  keyfile: None
  certfile: None
  ssl_version: 2
  cert_reqs: 0
  ca_certs: None
  suppress_ragged_eofs: True
  do_handshake_on_connect: False
  ciphers: None
  raw_paste_global_conf: []
  strip_header_spaces: False
```

</details>

<details>
<summary>
Printed settings in Gunicorn 20.1.0
</summary>

```
access_log_format                 = %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
accesslog                         = -
backlog                           = 2048
bind                              = ['0.0.0.0:5000']
ca_certs                          = None
capture_output                    = False
cert_reqs                         = 0
certfile                          = None
chdir                             = /Users/username/Workspace/tmp/test_gunicorn/test_print
check_config                      = False
child_exit                        = <ChildExit.child_exit()>
ciphers                           = None
config                            = gunicorn.conf.py
daemon                            = False
default_proc_name                 = app:app
disable_redirect_access_to_syslog = False
do_handshake_on_connect           = False
dogstatsd_tags                    =
enable_stdio_inheritance          = False
errorlog                          = -
forwarded_allow_ips               = ['127.0.0.1']
graceful_timeout                  = 30
group                             = 20
initgroups                        = False
keepalive                         = 2
keyfile                           = None
limit_request_field_size          = 8190
limit_request_fields              = 100
limit_request_line                = 4094
logconfig                         = None
logconfig_dict                    = {}
logger_class                      = gunicorn.glogging.Logger
loglevel                          = debug
max_requests                      = 0
max_requests_jitter               = 0
nworkers_changed                  = <NumWorkersChanged.nworkers_changed()>
on_exit                           = <OnExit.on_exit()>
on_reload                         = <OnReload.on_reload()>
on_starting                       = <OnStarting.on_starting()>
paste                             = None
pidfile                           = None
post_fork                         = <Postfork.post_fork()>
post_request                      = <PostRequest.post_request()>
post_worker_init                  = <PostWorkerInit.post_worker_init()>
pre_exec                          = <PreExec.pre_exec()>
pre_fork                          = <Prefork.pre_fork()>
pre_request                       = <PreRequest.pre_request()>
preload_app                       = False
print_config                      = True
proc_name                         = None
proxy_allow_ips                   = ['127.0.0.1']
proxy_protocol                    = False
pythonpath                        = None
raw_env                           = []
raw_paste_global_conf             = []
reload                            = False
reload_engine                     = auto
reload_extra_files                = []
reuse_port                        = False
secure_scheme_headers             = {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
sendfile                          = None
spew                              = False
ssl_version                       = 2
statsd_host                       = None
statsd_prefix                     =
strip_header_spaces               = False
suppress_ragged_eofs              = True
syslog                            = False
syslog_addr                       = unix:///var/run/syslog
syslog_facility                   = user
syslog_prefix                     = None
threads                           = 17
timeout                           = 240
tmp_upload_dir                    = None
umask                             = 0
user                              = 501
when_ready                        = <WhenReady.when_ready()>
worker_abort                      = <WorkerAbort.worker_abort()>
worker_class                      = sync
worker_connections                = 1000
worker_exit                       = <WorkerExit.worker_exit()>
worker_int                        = <WorkerInt.worker_int()>
worker_tmp_dir                    = None
workers                           = 17
wsgi_app                          = None
```

</details>

**Comment by wjentner** [comment]
Oh wow. I guess I just added it to the config thinking it works similarly to the other CLI flags and as it indeed printed the config (in the logs) I assumed it comes from the config option.

Thank you for the clarification!
